lib/codecs/src/decoding/format/vrl.rs

```diff
@@ -0,0 +1,257 @@
+#[derivative(Default)]
+pub struct VrlDeserializerOptions {
+    /// The [Vector Remap Language][vrl] (VRL) program to execute for each event.
+    ///
+    /// [vrl]: https://vector.dev/docs/reference/vrl
+    pub source: String,
+
+    /// The name of the timezone to apply to timestamp conversions that do not contain an explicit
+    /// time zone. The time zone name may be any name in the [TZ database][tz_database], or `local`
+    /// to indicate system local time.
+    ///
+    /// If not set, `local` will be used.
+    ///
+    /// [tz_database]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
+    #[serde(default)]
+    #[configurable(metadata(docs::advanced))]
+    pub timezone: Option<TimeZone>,
+}
+
+impl VrlDeserializerConfig {
+    /// Build the `VrlDeserializer` from this configuration.
+    pub fn build(&self) -> vector_common::Result<VrlDeserializer> {
+        let state = TypeState {
+            local: Default::default(),
+            external: ExternalEnv::default(),
+        };
+
+        match compile_vrl(
+            &self.vrl.source,
+            &vrl::stdlib::all(),
+            &state,
+            CompileConfig::default(),
+        ) {
+            Ok(result) => Ok(VrlDeserializer {
+                program: result.program,
+                timezone: self.vrl.timezone.unwrap_or(TimeZone::Local),
+            }),
+            Err(diagnostics) => Err(Formatter::new(&self.vrl.source, diagnostics)
+                .to_string()
+                .into()),
+        }
+    }
+
+    /// Return the type of event build by this deserializer.
+    pub fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema produced by the deserializer.
+    pub fn schema_definition(&self, log_namespace: LogNamespace) -> schema::Definition {
+        match log_namespace {
+            LogNamespace::Legacy => {
+                schema::Definition::empty_legacy_namespace().unknown_fields(Kind::any())
+            }
+            LogNamespace::Vector => {
+                schema::Definition::new_with_default_metadata(Kind::any(), [log_namespace])
+            }
+        }
+    }
+}
+
+/// Deserializer that builds `Event`s from a byte frame containing logs compatible with VRL.
+#[derive(Debug, Clone)]
+pub struct VrlDeserializer {
+    program: Program,
+    timezone: TimeZone,
+}
+
+fn parse_bytes(bytes: Bytes, log_namespace: LogNamespace) -> Event {
+    let bytes_deserializer = BytesDeserializerConfig::new().build();
+    let log_event = bytes_deserializer.parse_single(bytes, log_namespace);
+    Event::from(log_event)
+}
+
+impl Deserializer for VrlDeserializer {
+    fn parse(
+        &self,
+        bytes: Bytes,
+        log_namespace: LogNamespace,
+    ) -> vector_common::Result<SmallVec<[Event; 1]>> {
+        let event = parse_bytes(bytes, log_namespace);
+        match self.run_vrl(event, log_namespace) {
+            Ok(event) => Ok(smallvec![event]),
+            Err(e) => Err(e),
+        }
```

Comment:
nit: This definitely is simpler as a `map`:
```suggestion
        self.run_vrl(event, log_namespace).map(|event| smallvec![event])
```